python/dgl/runtime/ir/executor.py

```diff
@@ -1169,3 +995,254 @@ def CLEAR_FRAME_(fd):
+    lhs_map : var.Var
+        Variable for mapping lambda. The lambda returns the lhs id mapping
+        array on given context
+    rhs_map : var.Var
+        Variable for mapping lambda. The lambda returns the rhs id mapping
+        array on given context
+    out_map : var.Var
+        Variable for mapping lambda. The lambda returns the output id mapping
+        array on given context
+    ret : var.Var, optional
+        Variable for the result. If not give, a new variable will be created.
+
+    Returns
+    -------
+    var.Var
+        Variable for the result.
+    """
+    reg = IR_REGISTRY[OpCode.BINARY_REDUCE]
+    ret = var.new(reg['ret_type']) if ret is None else ret
+    get_current_prog().issue(reg['executor_cls'](
+        reducer, binary_op, graph, lhs, rhs, lhs_data, rhs_data, out_size,
+        lhs_map, rhs_map, out_map, ret))
+    return ret
+
+
+class CopyReduceExecutor(Executor):
+    """Executor for COPY_REDUCE
+
+    Parameters
+    ----------
+    reducer : str
+        String representing reduction to perform, can be "sum", "max", "min",
+        "mean", "prod", "none" (no reduction)
+    graph : var.Var
+        Variable for graph index lambda. The lambda returns the immutable graph
+        index given a context object.
+    target: int
+        The input target (src, dst, edge)
+    in_data : var.Var
+        Variable for the input data
+    out_size : int
+        Output size
+    in_map : var.Var
+        Variable for mapping lambda. The lambda returns the input id mapping
+        array on given context
+    out_map : var.Var
+        Variable for mapping lambda. The lambda returns the output id mapping
+        array on given context
+    ret : var.Var
+        Variable for the result.
+    """
+    def __init__(self, reducer, graph, target, in_data, out_size, in_map,
+                 out_map, ret):
+        self.reducer = reducer
+        self.graph = graph
+        self.target = target
+        self.in_data = in_data
+        self.out_size = out_size
+        self.in_map = in_map
+        self.out_map = out_map
+        self.ret = ret
+
+    def opcode(self):
+        return OpCode.COPY_REDUCE
+
+    def arg_vars(self):
+        return [self.reducer, self.graph, self.target, self.in_data,
+                self.out_size, self.in_map, self.out_map]
+
+    def ret_var(self):
+        return self.ret
+
+    def run(self):
+        in_data = self.in_data.data
+        ctx = utils.to_dgl_context(F.context(in_data))
+        graph = self.graph.data(ctx)
+        in_map = self.in_map.data(ctx) if self.in_map.data else None
+        out_map = self.out_map.data(ctx) if self.out_map.data else None
+        if not isinstance(in_map, tuple):
+            in_map = (in_map, in_map)
+        if not isinstance(out_map, tuple):
+            out_map = (out_map, out_map)
+        self.ret.data = F.copy_reduce(
+            self.reducer, graph, self.target, in_data, self.out_size, in_map,
+            out_map)
```

Comment:
Did we ever benchmark our binary/copy reducer against PyTorch SPMM?  Just curious.  We don't have to do it.

Reply:
We did. Our copy_reduce and torch.spmm both use cusparse, but our backward is more efficient since we cache the reverse csr. Compared with scatter, the result in our blogpost applies. PyG essentially used torch's scatter_add.

Reply:
Not yet. Will do for sure.